drivers/soundwire/intel_init.c

```diff
@@ -263,6 +269,25 @@ static struct sdw_intel_ctx
 		md->driver->probe(md, link);
 
 		list_add_tail(&link->list, &ctx->link_list);
+		bus = &link->cdns->bus;
+		/* Calculate number of slaves */
+		list_for_each_entry(slave, &bus->slaves, node)
+			num_slaves++;
+	}
+
+	ctx->ids = kcalloc(num_slaves, sizeof(*ctx->ids), GFP_KERNEL);
+	if (!ctx->ids)
+		goto err;
+
+	ctx->num_slaves = num_slaves;
+	i = 0;
+	list_for_each_entry(link, &ctx->link_list, list) {
+		bus = &link->cdns->bus;
+		list_for_each_entry(slave, &bus->slaves, node) {
+			ctx->ids[i].id = slave->id;
+			ctx->ids[i].link_id = bus->link_id;
+			i++;
```

Comment:
there is a chance of a future breakage. We counted the total number of elements of `ids[]` that we need above and allocated it accordingly. Now we just walk the lists without checking again. It's correct at the moment, but there is a chance a future change might break that. We can either add a check to an equivalent of `BUG_ON(i >= num_slaves)` or at least add a comment, that the array is allocated correctly above, we know we cannot overrun it...